### PR TITLE
Fix NtGdiGetTextFaceW to report face name correctly

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

This Pull Request fixes GetTextFace function to report the font face name correctly.
JIRA issue: [CORE-14071](https://jira.reactos.org/browse/CORE-14071)

## Proposed changes

- Initialize TEXTOBJ.FaceName member in TextIntRealizeFont function in "win32ss/gdi/ntgdi/freetype.c".
- Set the face name in NtGdiGetTextFaceW function by using TEXTOBJ.FaceName in "win32ss/gdi/ntgdi/text.c".

### CODE REVIEW

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -4518,8 +4519,20 @@ TextIntRealizeFont(HFONT FontHandle, PTEXTOBJ pTextObj)
...
+        if (NT_SUCCESS(Status))
+        {
+            RtlCopyMemory(TextObj->FaceName, NameW.Buffer, NameW.Length);
+            TextObj->FaceName[NameW.Length / sizeof(WCHAR)] = UNICODE_NULL;
+            RtlFreeUnicodeString(&NameW);
+        }
+
         Face = FontGdi->SharedFace->Face;
 
+
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/freetype.c]: Extra newline unneeded.

**Comment by yagoulas** [comment]
From what I understand LFONT::FullName, LFONT::Style and LFONT::FaceName are not used atm, correct? That is why using FaceName should be fine. A question that I have is if NtGdiGetTextFaceW always ends up getting the name of the realized font. From the look of it, this is correct as well. 

One small improvement would be not to let fLen become larger than LF_FACESIZE since this is the size of FaceName. And a last thing, would be to make sure that FaceName is always initialized even when the font isn't realized so that we avoid any catastrophic results.

A final note is that FaceName is one of the things that in the end should be moved to a new RFONT (which doesn't exist yet and the conversation to create it is a long one and doesn't belong in this discussion).

I think that even without my two suggestions it should work fine but it would be better to be extra sure that nothing wrong will happen.

File: win32ss/gdi/ntgdi/freetype.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

**Comment by katahiromz** [comment]
Yes, FullName, Style and FaceName were unused members of LFONT or TEXTOBJ.
Checking length against LF_FACESIZE is better idea.

File: win32ss/gdi/ntgdi/text.c [hunk 2]
```diff
@@ -525,7 +527,7 @@ NtGdiGetTextFaceW(
             return 0;
         }
         /* Terminate if we copied only part of the font name */
-        if (Count > 0 && Count < fLen)
+        if (0 < Count && Count <= fLen)
         {
             FaceName[Count - 1] = '\0';
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/text.c]: Naive question: We don't need SEH here?

Comment (yagoulas) [reply win32ss/gdi/ntgdi/text.c]: MmCopyToCaller does it.

File: win32ss/gdi/ntgdi/text.c [hunk 3]
```diff
@@ -3,7 +3,7 @@
  * LICENSE:         GPL - See COPYING in the top level directory
  * FILE:            win32ss/gdi/ntgdi/text.c
  * PURPOSE:         Text/Font
- * PROGRAMMER:
+ * PROGRAMMER:      Katayama Hirofumi MZ
```
Comment (ThFabba) [inline win32ss/gdi/ntgdi/text.c]: I'd recommend looking at the history and adding the other authors as well, this looks odd otherwise.

File: win32ss/gdi/ntgdi/text.c (hunk 2 above)
Comment (ThFabba) [reply win32ss/gdi/ntgdi/text.c]: Yeah except you need SEH every time you access user memory, because another thread could be deallocating it concurrently. So yes, we need SEH here.

File: win32ss/gdi/ntgdi/text.c [hunk 4]
```diff
@@ -512,20 +512,22 @@ NtGdiGetTextFaceW(
...
+        Status = MmCopyToCaller(FaceName, TextObj->FaceName, Count * sizeof(WCHAR));
         if (!NT_SUCCESS(Status))
         {
             TEXTOBJ_UnlockText(TextObj);
             SetLastNtError(Status);
             return 0;
         }
         /* Terminate if we copied only part of the font name */
-        if (Count > 0 && Count < fLen)
+        if (0 < Count && Count <= fLen)
```
Comment (ThFabba) [inline win32ss/gdi/ntgdi/text.c]: `Count > 0` to `0 < Count` is an unnecessary change, please revert.

Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: OK.

File: win32ss/gdi/ntgdi/text.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: How to do?

File: win32ss/gdi/ntgdi/text.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: I agree.

Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: https://git.reactos.org/?p=reactos-deprecated-gitsvn-dont-use.git;a=history;f=reactos/win32ss/gdi/ntgdi/text.c;h=b4130b6b2fc0423a13824e794122179ca85492cd;hb=refs/heads/master
The other authors were akhaldi, hbelusca, jimtabor, and tkreuzer.

File: win32ss/gdi/ntgdi/text.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: OK, SEH handled.

File: win32ss/gdi/ntgdi/text.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: OK, I added the authors.

File: win32ss/gdi/ntgdi/text.c [hunk 5]
```diff
@@ -512,24 +516,34 @@ NtGdiGetTextFaceW(
...
-            FaceName[Count - 1] = '\0';
+            _SEH2_TRY
+            {
+                FaceName[Count - 1] = '\0';
+            }
+            _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+            {
+                ret = 0;
+            }
+            _SEH2_END
```
Comment (ThFabba) [inline win32ss/gdi/ntgdi/text.c]: SEH code looks good, but please add a semicolon after _SEH2_END

Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: OK.

**Comment by yagoulas** [comment]
I think this is good to go. Do you guys have any objection?

**Comment by HBelusca** [comment]
LGTM to me; the full thing needs to be squashed however :)

File: win32ss/gdi/ntgdi/text.c [hunk 6]
```diff
@@ -3,7 +3,11 @@
  * LICENSE:         GPL - See COPYING in the top level directory
  * FILE:            win32ss/gdi/ntgdi/text.c
  * PURPOSE:         Text/Font
- * PROGRAMMER:
+ * PROGRAMMERS:     Amine Khaldi <amine.khaldi@reactos.org>
+ *                  Hermes Belusca-Maito <hermes.belusca@sfr.fr>
+ *                  Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
```
Comment (binarymaster) [inline win32ss/gdi/ntgdi/text.c]: Does this list is sorted alphabetically? If so, James should be above you.

Comment (HBelusca) [reply win32ss/gdi/ntgdi/text.c]: Usually no; this is added chronologically.

Comment (katahiromz) [reply win32ss/gdi/ntgdi/text.c]: Sorted the authors chronologically.

**Comment by JoachimHenze** [comment]
Unfortunately the acceptance of that pull request led to the regression of CORE-14170. Maybe some of the reviewers can help Katayama to figure out, what went wrong here.
